controllers/tasks.js

```diff
@@ -142,23 +143,59 @@ const updateTaskStatus = async (req, res, next) => {
   try {
     const taskId = req.params.id;
     const { dev } = req.query;
-    const { id: userId } = req.userData;
+    const { id: userId, username } = req.userData;
     const task = await tasks.fetchSelfTask(taskId, userId);
 
     if (task.taskNotFound) return res.boom.notFound("Task doesn't exist");
     if (task.notAssignedToYou) return res.boom.forbidden("This task is not assigned to you");
-    if (task.taskData.status === "VERIFIED")
+    if (task.taskData.status === "VERIFIED" || req.body.status === "MERGED")
       return res.boom.forbidden("Status cannot be updated. Please contact admin.");
 
-    await tasks.updateTask(req.body, taskId);
+    if (task.taskData.status === "COMPLETED" && req.body.percentCompleted < 100) {
+      if (req.body.status === "COMPLETED" || !req.body.status) {
+        return res.boom.forbidden("Task percentCompleted can't updated as status is COMPLETED");
+      }
+    }
+
+    if (req.body.status === "COMPLETED" && task.taskData.percentCompleted !== 100) {
+      if (req.body.percentCompleted !== 100) {
+        return res.boom.forbidden("Status cannot be updated. Task is not completed yet");
+      }
+    }
+
+    const taskLog = {
+      type: "task",
+      meta: {
+        userId,
+        taskId,
+        username,
+      },
+      body: {
+        subType: "update",
+        new: {},
+      },
+    };
+
+    if (req.body.status) {
+      taskLog.body.new.status = req.body.status;
+    }
+    if (req.body.percentCompleted) {
+      taskLog.body.new.percentCompleted = req.body.percentCompleted;
+    }
```

Comment:
In case of `req.body.status` & `req.body.percentCompleted` not being present then `taskLog.body.new.status` & `taskLog.body.new.percentCompleted` will be undefined, is this okay ?

Reply:
There are if statements for that reason only.

Reply:
What I meant was are `taskLog.boy.new.status` & `taskLog.body.new.percentCompleted` optional fields ? Is it fine okay if we do not have them ?

Reply:
no, it wouldn't be as the approach was to log the task event in a manner that the raw data should be self-explanatory about what was the event that happened with specificity